### PR TITLE
Create a BetterDrawerLayout component

### PR DESCRIPTION
**Pull request by j-piasecki**
## Description

Create a new BetterDrawerLayout component built with the new API of the gesture handler and Reanimated 2. It is supposed to be an almost drop-in replacement for the previous DrawerLayout.

## Test plan

Added a new example in the Example app.

### CODE REVIEW

File: src/components/BetterDrawerLayout.tsx [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+        translation = 0;
+      } else {
+        translation = containerWidth - drawerWidth;
+      }
+    }
+
+    // if the drawer is not visible move it off the screen to prevent it
+    // from intercepting touch events on Android
+    if (drawerOffset.value === 0) {
+      translation = 10000;
```
Comment (jakub-gonet) [inline src/components/BetterDrawerLayout.tsx:514]: If you go with `pointerEvents` then you have to change the logic here as well.

File: src/components/BetterDrawerLayout.tsx [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+    }
+
+    return {
+      flexDirection: reverseContentDirection ? 'row-reverse' : 'row',
+      transform: [{ translateX: translation }],
+    };
+  }, [drawerSlide, drawerWidth, containerWidth, fromLeft]);
+
+  const containerStyle = useAnimatedStyle(() => {
+    let translation = 0;
```
Comment (jakub-gonet) [inline src/components/BetterDrawerLayout.tsx:524]: If you name it `translateX` then you'll be able to use shorthand in returned styles.

File: src/components/BetterDrawerLayout.tsx [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+      // the container should be moved with the gesture
+      translation = drawerOffset.value;
+    }
+
+    return {
+      transform: [{ translateX: translation }],
+    };
+  }, [containerSlide]);
+
+  if (ref != null) {
```
Comment (jakub-gonet) [inline src/components/BetterDrawerLayout.tsx:536]:
```suggestion
  if (ref !== null) {
```

Comment (jakub-gonet) [reply src/components/BetterDrawerLayout.tsx:536]: If its type is `null | undefined` then leave it.

File: examples/Example/src/basic/betterHorizontalDrawer/index.tsx [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        style={styles.pageInput}
+        placeholder="Just an input field to test kb dismiss mode"
+      />
+    </View>
+  );
+}
+
+function DrawerContent(
+  offset: Animated.SharedValue<number>,
+  parralax: boolean,
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/basic/betterHorizontalDrawer/index.tsx]: Fixed in 4c0dd6c49fe6a14a5e07fbd1ebb537e76dd32144

File: src/components/BetterDrawerLayout.tsx [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+   * the drawer that is visible
+   */
+  onDrawerSlide?: (position: number) => void;
+
+  children?: React.ReactNode;
+}
+
+const IDLE: DrawerState = 'Idle';
+const DRAGGING: DrawerState = 'Dragging';
+const SETTLING: DrawerState = 'Settling';
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/components/BetterDrawerLayout.tsx]: Done in 4a0b2acdeb914d3cadec8c0b74ea274022e61df7, although [babel doesn't support `const enums`](https://babeljs.io/docs/en/babel-plugin-transform-typescript#caveats) so I made it to a normal enum.

File: src/components/BetterDrawerLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+          translateX:
+            // when the overlay should not be visible move it off the screen
+            // to prevent it from intercepting touch events on Android
+            props.drawerType !== 'front' || props.progress.value === 0
+              ? 10000
+              : 0,
+        },
+      ],
+    }),
+    [props]
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/components/BetterDrawerLayout.tsx]: Removed in dfb1b8655a139c4c1e9a514888cc77ce4c76dbdd

File: src/components/BetterDrawerLayout.tsx [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+  //default props
+  const drawerWidth = props.drawerWidth ?? 200;
+  const drawerPosition = props.drawerPosition ?? 'left';
+  const drawerType = props.drawerType ?? 'front';
+  const edgeWidth = props.edgeWidth ?? 20;
+  const minSwipeDistance = props.minSwipeDistance ?? 3;
+  const overlayColor = props.overlayColor ?? 'rgba(0, 0, 0, 0.7)';
+  const drawerLockMode = props.drawerLockMode ?? 'unlocked';
+  const enableTrackpadTwoFingerGesture =
+    props.enableTrackpadTwoFingerGesture ?? false;
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/components/BetterDrawerLayout.tsx]: Done in 8fe550dcf86063d07ec40ab498f6cd59129a5bc9

File: src/components/BetterDrawerLayout.tsx [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+    // this is required in addition to the similar call below, because the gesture
+    // doesn't change `drawerVisible` state to prevent re-render during gesture
+    // so when dragging from closed it wouldn't hide the status bar
+    if (props.hideStatusBar === true) {
+      StatusBar.setHidden(true, props.statusBarAnimation ?? 'slide');
+    }
+  }
+
+  function setState(newState: DrawerState, willShow: boolean) {
+    if (props.hideStatusBar === true) {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/components/BetterDrawerLayout.tsx]: Done in 4ac22a17b646573a931f2e69d95e6bf336efb016

File: src/components/BetterDrawerLayout.tsx (hunk 2 above)
Comment (j-piasecki) [reply src/components/BetterDrawerLayout.tsx]: Done in a2e465205ff661aa0f1f51c7e9064d9478016cb3

File: src/components/BetterDrawerLayout.tsx (hunk 3 above)
Comment (j-piasecki) [reply src/components/BetterDrawerLayout.tsx]: Changed it in 4c05343f1bc75bb5a3019e20421322641a5f754e

File: src/components/BetterDrawerLayout.tsx [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+
+  function onDragStart() {
+    if (props.keyboardDismissMode === 'on-drag') {
+      Keyboard.dismiss();
+    }
+
+    // this is required in addition to the similar call below, because the gesture
+    // doesn't change `drawerVisible` state to prevent re-render during gesture
+    // so when dragging from closed it wouldn't hide the status bar
+    if (props.hideStatusBar === true) {
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/components/BetterDrawerLayout.tsx]: Done in 4ac22a17b646573a931f2e69d95e6bf336efb016

File: examples/Example/src/basic/betterHorizontalDrawer/index.tsx [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+import {
+  DrawerLayoutController,
+  DrawerType,
+  RectButton,
+  BetterDrawerLayout,
+} from 'react-native-gesture-handler';
+import Animated, { useAnimatedStyle } from 'react-native-reanimated';
+
+const TYPES: DrawerType[] = ['front', 'back', 'back', 'slide'];
+const PARALLAX: boolean[] = [false, false, true, false];
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/basic/betterHorizontalDrawer/index.tsx:14]: What do you mean exactly?

File: src/components/BetterDrawerLayout.tsx [hunk 11]
```diff
@@ -0,0 +1,611 @@
...
+      backgroundColor: props.color,
+      opacity: props.progress.value,
+      transform: [
+        {
+          translateX:
+            // when the overlay should not be visible move it off the screen
+            // to prevent it from intercepting touch events on Android
+            props.drawerType !== 'front' || props.progress.value === 0
+              ? 10000
+              : 0,
```
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/components/BetterDrawerLayout.tsx]: It doesn't seem to work (or I'm doing something wrong) but the way it's now also hides the overlay in modes other than front. And I'm not sure if 10000 is enough but I think if a view is far enough from the screen it's not rendered at all improving performance, is that right?

File: examples/Example/src/basic/betterHorizontalDrawer/index.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (jakub-gonet) [reply examples/Example/src/basic/betterHorizontalDrawer/index.tsx:14]: `const PARALLAX = ...`

File: src/components/BetterDrawerLayout.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (jakub-gonet) [reply src/components/BetterDrawerLayout.tsx]: That's the case on Android for sure. You can also try `visibilty: hidden`.

File: examples/Example/src/basic/betterHorizontalDrawer/index.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (j-piasecki) [reply examples/Example/src/basic/betterHorizontalDrawer/index.tsx]: Removed in 67df4bff43a5798ef2de08abc06a1bae99425386

File: src/components/BetterDrawerLayout.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (j-piasecki) [reply src/components/BetterDrawerLayout.tsx]: Doesn't work with `visibility: hidden` either.

[the remark this answers is not in the document]
Comment (jakub-gonet) [reply src/components/BetterDrawerLayout.tsx]: Let's leave it then.

**Review by jakub-gonet — APPROVED** [review]

**Pull request merged**